src/box/lua/config/applier/credentials.lua

```diff
@@ -525,17 +525,93 @@ end
                           name, privs, obj_type, obj_name)
 
+    -- This key identifies an alert to be dropped on a successful
+    -- operation and to be set on an error.
+    --
+    -- Let's consider the following situation: a space does not
+    -- exist, but the configuration contains privileges for it
+    -- (for example 'read' and 'write'). The warning will be
+    -- issued with the "box.schema.user.grant(<...>, 'read,write',
+    -- <...>)" key.
+    --
+    -- Now, let's assume that the space is created. We should drop
+    -- the warning and, so, we should generate exactly same key to
+    -- pass to the :drop() aboard method.
+    --
+    -- 1. Is it possible that the key to drop contains just 'read',
+    --    not 'read,write'? Seems unlikely.
+    --
+    --    If the configuration was changed before the space
+    --    creation, then the old ('read,write') alert is dropped
+    --    together with all the alerts from the old configuration
+    --    and the new alert ('read') is issued on applying the new
+    --    configuration.
+    --
+    --    OTOH, if a space creation is in the same transaction
+    --    with granting a 'write' permission, it seems possible.
+    --    However, it looks unlikely that permissions for the same
+    --    user and the same object (space/function/sequence) are
+    --    granted using box.schema.user.grant() and using the
+    --    declarative configuration at the same time.
+    --
+    -- 2. Is it possible that separate calls with 'read' and
+    --    'write' will be performed? Seems, no.
+    --
+    --    privileges_from_config() glues all the config privileges
+    --    into a mapping, so different configuration forms do not
+    --    affect the result. The following two configurations are
+    --    equivalent.
+    --
+    --    credentials:
+    --      users:
+    --        myuser:
+    --          privileges:
+    --          - permissions: [read, write]
+    --            spaces: [s]
+    --
+    --    credentials:
+    --      users:
+    --        myuser:
+    --          privileges:
+    --          - permissions: [read]
+    --            spaces: [s]
+    --          - permissions: [write]
+    --            spaces: [s]
+    --
+    -- 3. Is it possible that 'write,read' is in the key to drop,
+    --    while 'read,write' is in the original alert? No.
+    --
+    --    Even if we got a different order as input of this
+    --    function, it is sorted before the key generation: see
+    --    table.sort() above.
+    --
+    -- What to do if we found an inconsistency that leads to one
+    -- of the three situations above?
+    --
+    -- The simplest solution is to issue several alerts for each
+    -- object: one for the 'read' permission, one for 'write'
+    -- permission and so on.
+    --
+    -- More complicated solution is to perform a search for
+    -- relevant alerts on successful privilege granting to modify
+    -- or drop a right alert.
```

Comment:
I don't see, how this approach is more complicated, the search in `_aboard:drop()` is still performed. I would say it is just less durable.

Reply:
It is a hash table lookup now (using the `key`). In the described scenario we need to perform a traverse and match the keys against a condition. Or allow to mark several alerts with one key/label and search using this label. I'll expand the comment.

Reply:
Oh, now I get it.

Reply:
I've expanded it a bit.

```diff
diff --git a/src/box/lua/config/applier/credentials.lua b/src/box/lua/config/applier/credentials.lua
index 62f92e2781..332444bbdc 100644
--- a/src/box/lua/config/applier/credentials.lua
+++ b/src/box/lua/config/applier/credentials.lua
@@ -601,9 +601,10 @@ local privileges_action_f = function(grant_or_revoke, role_or_user, name, privs,
     -- object: one for the 'read' permission, one for 'write'
     -- permission and so on.
     --
-    -- More complicated solution is to perform a search for
+    -- More complicated solution is to perform a traverse for
     -- relevant alerts on successful privilege granting to modify
-    -- or drop a right alert.
+    -- or drop a right alert. In this case we don't know an exact
+    -- list of privileges to form a key to drop the alert.
     local key = ('box.schema.%s.%s(%q, %q, %q, %q)'):format(
         role_or_user, grant_or_revoke, name, privs, obj_type, obj_name)
```

However, this comment will likely gone during the work on the problem with an object renaming.

Reply:
This comment is removed in the new implementation (see [below](https://github.com/tarantool/tarantool/pull/9598#discussion_r1457157778)).